### PR TITLE
[RLlib; docs] Docs do-over (new API stack): Env pages vol 02.

### PR DESCRIPTION
**Pull request by sven1977**
Do-over of RLlib docs (new API stack):
* Redo existing rllib-env.rst page.
* Add new multi-agent-envs.rst page (and move all the multi-agent relevant docs here, plus rewrite them)
* Add new hierarchical-envs.rst page (and move the paragraph on hierarchical envs here)
* Add new external-envs.rst page (and move the paragraphs describing these here)
* Add new example classes and scripts (also add to CI), highlighting the different multi-agent movement patterns: sequential vs simultaneous.
* Add new figures to docs.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/rllib/external-envs.rst [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+together with a dummy (CartPole) client providing a template for any external application or simulator.
+
+.. note::
+    External application support is still work-in-progress on RLlib's new API stack. The Ray team
+    is working on more examples for custom EnvRunner implementations (besides
+    `the already available tcp-based one <https://github.com/ray-project/ray/blob/master/rllib/env/tcp_client_inference_env_runner.py>`__)
+    as well as client-side, non-python RLlib-adapters, for example for popular game engines and other
+    simulation software.
+
+
```
Comment (sven1977) [inline doc/source/rllib/external-envs.rst:44]: Still working on the rest of this very file ... Everything else (all other files) is ready, though.

File: doc/source/rllib/external-envs.rst [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        "input": (
+            lambda ioctx: PolicyServerInput(ioctx, SERVER_ADDRESS, SERVER_PORT)
+        ),
+        # Use the existing algorithm process to run the server.
+        "num_env_runners": 0,
+    }
+
+Clients can then connect in either *local* or *remote* inference mode.
+In local inference mode, copies of the policy are downloaded from the server and cached on the client for a configurable period of time.
+This allows actions to be computed by the client without requiring a network round trip each time.
```
Comment (angelinalg) [inline doc/source/rllib/external-envs.rst:86]:
```suggestion
This approach allows actions to be computed by the client without requiring a network round trip each time.
```

File: doc/source/rllib/external-envs.rst [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    action = client.get_action(episode_id, cur_obs)
+    ...
+    client.end_episode(episode_id, last_obs)
+
+To understand the difference between standard envs, external envs, and connecting with a ``PolicyClient``, refer to the following figure:
+
+.. https://docs.google.com/drawings/d/1hJvT9bVGHVrGTbnCZK29BYQIcYNRbZ4Dr6FOPMJDjUs/edit
+.. image:: images/rllib-external.svg
+
+Try it yourself by launching either a
```
Comment (angelinalg) [inline doc/source/rllib/external-envs.rst:105]:
```suggestion
Try it by launching either a
```

File: doc/source/rllib/external-envs.rst [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+`simple CartPole server <https://github.com/ray-project/ray/blob/master/rllib/examples/envs/external_envs/cartpole_server.py>`__ (see below), and connecting it to any number of clients
+(`cartpole_client.py <https://github.com/ray-project/ray/blob/master/rllib/examples/envs/external_envs/cartpole_client.py>`__) or
+run a `Unity3D learning sever <https://github.com/ray-project/ray/blob/master/rllib/examples/envs/external_envs/unity3d_server.py>`__
```
Comment (angelinalg) [inline doc/source/rllib/external-envs.rst:108]:
```suggestion
running a `Unity3D learning sever <https://github.com/ray-project/ray/blob/master/rllib/examples/envs/external_envs/unity3d_server.py>`__
```

File: doc/source/rllib/hierarchical-envs.rst [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+.. include:: /_includes/rllib/new_api_stack.rst
+
+.. _rllib-hierarchical-environments-doc:
+
+
+Hierarchical Environments
+=========================
+
+Hierarchical training can be implemented as a special case of multi-agent RL. For example, consider a two-level hierarchy of policies,
```
Comment (angelinalg) [inline doc/source/rllib/hierarchical-envs.rst:11]:
```suggestion
You can implement hierarchical training as a special case of multi-agent RL. For example, consider a two-level hierarchy of policies,
```

File: doc/source/rllib/hierarchical-envs.rst [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+where a top-level policy issues high level tasks that are executed at a finer timescale by one or more low-level policies.
```
Comment (angelinalg) [inline doc/source/rllib/hierarchical-envs.rst:12]:
```suggestion
where a top-level policy issues high level tasks that RLlib executes at a finer timescale by one or more low-level policies.
```

File: doc/source/rllib/hierarchical-envs.rst [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+but the low-level agents wait for the high-level agent to issue an action, then act n times before handing
+back control to the high-level agent:
+
+.. code-block:: text
+
+   top-level: action_0 -----------------------------------> action_1 ->
+   low-level: ---------> action_0 -> action_1 -> action_2 ------------>
+
+
+Any of these hierarchical action patterns can be implemented as a multi-agent environment with various
```
Comment (angelinalg) [inline doc/source/rllib/hierarchical-envs.rst:30]:
```suggestion
You can implement any of these hierarchical action patterns as a multi-agent environment with various
```

File: doc/source/rllib/hierarchical-envs.rst [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+types of agents, for example a high-level agent and a low-level agent. When set up using the correct
+agent to module mapping functions, from RLlib's perspective, the problem boils down to a simple independent
```
Comment (angelinalg) [inline doc/source/rllib/hierarchical-envs.rst:32]:
```suggestion
agent to module mapping functions, from RLlib's perspective, the problem becomes a simple independent
```

File: doc/source/rllib/hierarchical-envs.rst [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+multi-agent problem with different types of policies.
+
+Your configuration might look something like this:
```
Comment (angelinalg) [inline doc/source/rllib/hierarchical-envs.rst:35]:
```suggestion
Your configuration might look something like the following:
```

File: doc/source/rllib/hierarchical-envs.rst [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+            policies_to_train=["top_level"],
+        )
+    )
+
+
+In this setup, the appropriate rewards at any hierarchy level should be provided by the multi-agent env implementation.
+The environment class is also responsible for routing between agents, for example conveying `goals <https://arxiv.org/pdf/1703.01161.pdf>`__ from higher-level
+agents to lower-level agents as part of the lower-level agent observation.
+
+See `here for a runnable example of a hierarchical env <https://github.com/ray-project/ray/blob/master/rllib/examples/hierarchical/hierarchical_training.py>`__.
```
Comment (angelinalg) [inline doc/source/rllib/hierarchical-envs.rst:57]:
```suggestion
See `this runnable example of a hierarchical env <https://github.com/ray-project/ray/blob/master/rllib/examples/hierarchical/hierarchical_training.py>`__.
```

File: doc/source/rllib/multi-agent-envs.rst [hunk 11]
```diff
@@ -0,0 +1,533 @@
...
+========================
+
+In a multi-agent environment, multiple "agents" act simultaneously, in a turn-based
+sequence, or through an arbitrary combination of both.
+
+For instance, in a traffic simulation, there might be multiple "car" and
+"traffic light" agents interacting simultaneously, whereas in a board game,
+two or more agents may act in a turn-based sequence.
+
+Several different policy networks may be used to control the various agents.
```
Comment (angelinalg) [inline doc/source/rllib/multi-agent-envs.rst:17]:
```suggestion
RLlib may use several different policy networks to control the various agents.
```

**Review by angelinalg — COMMENTED** [review]
Not done yet, but releasing some comments.

**Comment by angelinalg** [comment]
I only got as far as [doc/source/rllib/multi-agent-envs.rst](https://github.com/ray-project/ray/pull/48542/files/4ce05cc482633f06247806251e494402e79f9af0#diff-7f895481d6a75d3653f16bc5bd2f84374bde5695cb55bc71161d67a7eff5d994) but I will pick up again tomorrow.

File: doc/source/rllib/rllib-env.rst [hunk 12]
```diff
@@ -7,598 +7,300 @@
...
 
-Here are two ways to implement a centralized critic compatible with the multi-agent API:
+    Due to Ray's distributed nature, gymnasium's own registry is incompatible
+    with Ray. Always use the registration method documented here to ensure
+    remote Ray actors can access your custom environments.
 
-**Strategy 1: Sharing experiences in the trajectory preprocessor**:
+In the preceding example, the ``env_creator`` function takes a ``config`` argument.
+This config is primarily a dictionary containing required settings.
+However, you can also access additional properties within the ``config`` variable (ex.
```
Comment (angelinalg) [inline doc/source/rllib/rllib-env.rst:219]:
```suggestion
However, you can also access additional properties within the ``config`` variable. For example,
```

File: doc/source/rllib/rllib-env.rst [hunk 13]
```diff
@@ -7,598 +7,300 @@
...
+``config.worker_index`` to get the remote EnvRunner index or ``config.num_workers``
+for the total number of EnvRunners used). This can help customize environments
```
Comment (angelinalg) [inline doc/source/rllib/rllib-env.rst:221]:
```suggestion
for the total number of EnvRunners used. This approach can help customize environments
```

File: doc/source/rllib/rllib-env.rst [hunk 14]
```diff
@@ -7,598 +7,300 @@
...
-Here are two ways to implement a centralized critic compatible with the multi-agent API:
+    Due to Ray's distributed nature, gymnasium's own registry is incompatible
+    with Ray. Always use the registration method documented here to ensure
+    remote Ray actors can access your custom environments.
 
-**Strategy 1: Sharing experiences in the trajectory preprocessor**:
+In the preceding example, the ``env_creator`` function takes a ``config`` argument.
+This config is primarily a dictionary containing required settings.
+However, you can also access additional properties within the ``config`` variable (ex.
+``config.worker_index`` to get the remote EnvRunner index or ``config.num_workers``
```
Comment (angelinalg) [inline doc/source/rllib/rllib-env.rst:220]:
```suggestion
use ``config.worker_index`` to get the remote EnvRunner index or ``config.num_workers``
```

File: doc/source/rllib/rllib-env.rst [hunk 15]
```diff
@@ -7,598 +7,300 @@
...
-   top_level ---------------------------------------------------------------> top_level --->
-   mid_level_0 -------------------------------> mid_level_0 ----------------> mid_level_1 ->
-   low_level_0 -> low_level_0 -> low_level_0 -> low_level_1 -> low_level_1 -> low_level_2 ->
-
-This can be implemented as a multi-agent environment with three types of agents. Each higher-level action creates a new lower-level agent instance with a new id (e.g., ``low_level_0``, ``low_level_1``, ``low_level_2`` in the above example). These lower-level agents pop in existence at the start of higher-level steps, and terminate when their higher-level action ends. Their experiences are aggregated by policy, so from RLlib's perspective it's just optimizing three different types of policies. The configuration might look something like this:
-
-.. code-block:: python
+    class EnvDependingOnWorkerAndVectorIndex(gym.Env):
+        def __init__(self, config):
+            # pick actual env based on worker and env indexes
```
Comment (angelinalg) [inline doc/source/rllib/rllib-env.rst:231]:
```suggestion
            # Pick actual env based on worker and env indexes.
```

File: doc/source/rllib/rllib-env.rst [hunk 16]
```diff
@@ -7,598 +7,300 @@
...
-agents to lower-level agents as part of the lower-level agent observation.
+    register_env("multi_env", lambda config: MultiEnv(config))
 
-See this file for a runnable example: `hierarchical_training.py <https://github.com/ray-project/ray/blob/master/rllib/examples/hierarchical/hierarchical_training.py>`__.
+.. tip::
 
-External Agents and Applications
---------------------------------
+    When using logging within an environment, the configuration must be done
+    inside the environment (running within Ray workers). Pre-Ray logging
```
Comment (angelinalg) [inline doc/source/rllib/rllib-env.rst:249]:
```suggestion
    inside the environment (running within Ray workers). RLlib ignores pre-Ray logging
```

File: doc/source/rllib/rllib-env.rst [hunk 17]
```diff
@@ -7,598 +7,300 @@
...
+    configurations will be ignored. Use the following code to connect to Ray's
```
Comment (angelinalg) [inline doc/source/rllib/rllib-env.rst:250]:
```suggestion
    configurations. Use the following code to connect to Ray's
```

File: doc/source/rllib/rllib-env.rst [hunk 18]
```diff
@@ -7,598 +7,300 @@
...
-and `n client(s) <https://github.com/ray-project/ray/blob/master/rllib/examples/envs/external_envs/cartpole_client.py>`__
-scripts, in which we setup an RLlib policy server that listens on one or more ports for client connections
-and connect several clients to this server to learn the env.
+.. figure:: images/envs/env_runners.svg
+    :width: 600
 
-Another `example <https://github.com/ray-project/ray/blob/master/rllib/examples/envs/external_envs/unity3d_server.py>`__ shows,
-how to run a similar setup against a Unity3D external game engine.
+    **EnvRunner with gym.Env setup:** Environments in RLlib are located within the :py:class:`~ray.rllib.envs.env_runner.EnvRunner` actors, whose number
+    (`n`) you can scale through the `config.env_runners(num_env_runners=..)` setting. Each such :py:class:`~ray.rllib.envs.env_runner.EnvRunner` actor
```
Comment (angelinalg) [inline doc/source/rllib/rllib-env.rst:266]:
```suggestion
    (`n`) you can scale through the `config.env_runners(num_env_runners=..)` setting. Each :py:class:`~ray.rllib.envs.env_runner.EnvRunner` actor
```

File: doc/source/rllib/rllib-env.rst [hunk 19]
```diff
@@ -7,598 +7,300 @@
...
+    itself can hold more than one `gymnasium <https://gymnasium.farama.org>`__ environment (vectorized). You can set the number
```
Comment (angelinalg) [inline doc/source/rllib/rllib-env.rst:267]:
```suggestion
    can hold more than one `gymnasium <https://gymnasium.farama.org>`__ environment (vectorized). You can set the number
```

File: doc/source/rllib/rllib-env.rst [hunk 20]
```diff
@@ -7,598 +7,300 @@
...
+    of individual environment copies per EnvRunner through `config.env_runners(num_envs_per_env_runner=..)`.
 
 
-Logging off-policy actions
-~~~~~~~~~~~~~~~~~~~~~~~~~~
+There are two methods to scale sample collection with RLlib and `gymnasium <https://gymnasium.farama.org>`__ environments, both of which
```
Comment (angelinalg) [inline doc/source/rllib/rllib-env.rst:271]:
```suggestion
There are two methods to scale sample collection with RLlib and `gymnasium <https://gymnasium.farama.org>`__ environments. You can use both
```

File: doc/source/rllib/rllib-env.rst [hunk 21]
```diff
@@ -7,598 +7,300 @@
...
+can be used in combination.
```
Comment (angelinalg) [inline doc/source/rllib/rllib-env.rst:272]:
```suggestion
in combination.
```

File: doc/source/rllib/rllib-env.rst [hunk 22]
```diff
@@ -7,598 +7,300 @@
...
 
-ExternalEnv provides a ``self.log_action()`` call to support off-policy actions. This allows the client to make independent decisions, e.g., to compare two different policies, and for RLlib to still learn from those off-policy actions. Note that this requires the algorithm used to support learning from off-policy decisions (e.g., DQN).
+1. **Distribute across multiple processes:** RLlib creates multiple
+   :py:class:`~ray.rllib.envs.env_runner.EnvRunner` instances (each a Ray actor) for experience collection,
```
Comment (angelinalg) [inline doc/source/rllib/rllib-env.rst:275]:
```suggestion
   :py:class:`~ray.rllib.envs.env_runner.EnvRunner` instances, each a Ray actor, for experience collection,
```

File: doc/source/rllib/rllib-env.rst [hunk 23]
```diff
@@ -7,598 +7,300 @@
...
+   controlled through your :py:class:`~ray.rllib.algorithms.algorithm_config.AlgorithmConfig`:
+   ``config.env_runners(num_env_runners=..)``.
 
-.. seealso::
+1. **Vectorization within a single process:** Many environments achieve high
+   frame rates per core but are limited by policy inference latency. To address
+   this, create multiple environments per process and thus batch the policy forward pass
```
Comment (angelinalg) [inline doc/source/rllib/rllib-env.rst:281]:
```suggestion
   this limitation, create multiple environments per process to batch the policy forward pass
```

File: doc/source/rllib/rllib-env.rst [hunk 24]
```diff
@@ -7,598 +7,300 @@
...
+   across these vectorized environments. Set ``config.env_runners(num_envs_per_env_runner=..)``
+   to create more than one environment copy per :py:class:`~ray.rllib.envs.env_runner.EnvRunner`
+   actor. Additionally, you can make the individual sub-environments within a vector
+   independent processes (through python's multiprocessing used by gymnasium).
```
Comment (angelinalg) [inline doc/source/rllib/rllib-env.rst:285]:
```suggestion
   independent processes through Python's multiprocessing used by gymnasium.
```

File: doc/source/rllib/rllib-env.rst [hunk 25]
```diff
@@ -7,598 +7,300 @@
...
+   Set `config.env_runners(remote_worker_envs=True)` to create individual subenvironments as separate processes
+   and step them in parallel.
 
-    `Offline Datasets <rllib-offline.html>`__ provide higher-level interfaces for working with off-policy experience datasets.
+.. note::
 
-External Application Clients
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    Multi-agent setups are not vectorizable yet. The Ray team is working on a solution for
```
Comment (angelinalg) [inline doc/source/rllib/rllib-env.rst:291]:
```suggestion
    Multi-agent setups aren't vectorizable yet. The Ray team is working on a solution for
```

File: doc/source/rllib/rllib-env.rst [hunk 26]
```diff
@@ -7,598 +7,300 @@
...
+    this restriction by utilizing `gymnasium >= 1.x` custom vectorization feature.
```
Comment (angelinalg) [inline doc/source/rllib/rllib-env.rst:292]:
```suggestion
    this restriction by using the `gymnasium >= 1.x` custom vectorization feature.
```

File: doc/source/rllib/rllib-env.rst [hunk 27]
```diff
@@ -7,598 +7,300 @@
...
-
-Advanced Integrations
----------------------
-
-For more complex / high-performance environment integrations, you can instead extend the low-level
-`BaseEnv <https://github.com/ray-project/ray/blob/master/rllib/env/base_env.py>`__ class.
-This low-level API models multiple agents executing asynchronously in multiple environments.
-A call to ``BaseEnv:poll()`` returns observations from ready agents keyed by 1) their environment, then 2) agent ids.
-Actions for those agents are sent back via ``BaseEnv:send_actions()``. BaseEnv is used to implement all the other env types in RLlib, so it offers a superset of their functionality.
+Some environments may require substantial resources to initialize and run. Should your environments require
```
Comment (angelinalg) [inline doc/source/rllib/rllib-env.rst:303]:
```suggestion
Some environments may require substantial resources to initialize and run. If your environments require
```

**Review by angelinalg — APPROVED** [review]
Good work on all this writing. If this is a good time to change titles to sentence case, that would be more consistent with our style guide. Otherwise, don't worry about it. It's definitely not a blocker. Hope the suggestions are helpful. Will approve to not block you. Sorry for the delay.

File: python/ray/data/_internal/planner/plan_udf_map_op.py [hunk 28]
```diff
@@ -46,7 +46,7 @@
 )
 from ray.data.context import DataContext
 from ray.data.exceptions import UserCodeException
-from ray.util.rpdb import _is_ray_debugger_post_mortem_enabled
+#from ray.util.rpdb import _is_ray_debugger_post_mortem_enabled
```
Comment (sven1977) [inline python/ray/data/_internal/planner/plan_udf_map_op.py:49]:
```suggestion
from ray.util.rpdb import _is_ray_debugger_post_mortem_enabled
```

File: python/ray/data/exceptions.py [hunk 29]
```diff
@@ -6,7 +6,7 @@
 from ray.exceptions import UserCodeException
 from ray.util import log_once
 from ray.util.annotations import DeveloperAPI
-from ray.util.rpdb import _is_ray_debugger_post_mortem_enabled
+#from ray.util.rpdb import _is_ray_debugger_post_mortem_enabled
```
Comment (sven1977) [inline python/ray/data/exceptions.py:9]:
```suggestion
from ray.util.rpdb import _is_ray_debugger_post_mortem_enabled
```

**Review by simonsays1980 — APPROVED** [review]
LGTM. Great overview about environments and detailed description of the delicate multi-agent environments and episodes users have to take care of.

File: doc/source/rllib/external-envs.rst [hunk 30]
```diff
@@ -0,0 +1,192 @@
...
+    connects to RLlib, which runs as a server through a tcp-cabable, custom EnvRunner.
+    The simulator sends batches of data from time to time to the server and in turn receives weights updates.
+    For better performance, actions are computed locally on the client side.
+
+.. todo (sven): show new image here with UE5
+.. .. figure:: images/rllib-training-inside-a-unity3d-env.png
+.. scale: 75 %
+..    A Unity3D soccer game being learnt by RLlib via the ExternalEnv API.
+
+RLlib provides an `external messaging protocol <https://github.com/ray-project/ray/blob/master/rllib/env/utils/external_env_protocol.py>`__
```
[the remark this answers is not in the document]
Comment (sven1977) [reply doc/source/rllib/external-envs.rst:32]: Yeah, let's make this a widely adopted standard! :)

File: doc/source/rllib/external-envs.rst [hunk 31]
```diff
@@ -0,0 +1,192 @@
...
+    as well as various client-side, non-python RLlib-adapters, for example for popular game engines and other
+    simulation software.
+
+
+.. _rllink-protocol-docs:
+
+The RLlink Protocol
+-------------------
+
+RLlink is a simple, stateful protocol designed for communication between a reinforcement learning (RL) server (ex., RLlib) and an
```
[the remark this answers is not in the document]
Comment (sven1977) [reply doc/source/rllib/external-envs.rst:52]: Definitely in the next iteration! Trying to keep it as simple as possible for this very first iteration. For now, this is just about the message types (what to say when and what to expect back from server?), not really the actual implementation of the messages.

**Pull request merged**